docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,311 @@
+
+As an AKS user looking to adopt Cluster API over an existing infrastructure managed solution, I want to be
+able to use all AKS features natively from CAPZ so that I can adopt Cluster API with the confidence that all
+the AKS features I currently utilize are still supported.
+
+### API Design Options
+
+There are a few different ways the entire AKS API surface area could be exposed from the CAPZ API. The
+following options all rely on ASO's ManagedCluster and ManagedClustersAgentPool resources to define the full
+AKS API, roughly organized in order of increasing responsibility for CAPZ. The examples below use
+AzureManagedControlPlane and ManagedCluster to help illustrate, but all of the same ideas should also apply to
+AzureManagedMachinePool and ManagedClustersAgentPool.
+
+#### Option 1: CAPZ resource references an existing ASO resource
+
+Here, the AzureManagedControlPlane spec would be updated with a field that references an ASO ManagedCluster
+resource:
+
+```go
+type AzureManagedControlPlaneSpec struct {
+	...
+	
+	// ManagedClusterName is the name of the ASO ManagedCluster backing this AzureManagedControlPlane.
+	ManagedClusterName string `json:"managedClusterName"`
+}
+```
+
+CAPZ will _not_ create this ManagedCluster and instead rely on it being created by any other means. CAPZ's
+`aks` flavor template will be updated to include a ManagedCluster to fulfill this requirement. CAPZ will also
+not modify the ManagedCluster except to fulfill CAPI contracts, such as managing replica count. Users modify
+other parameters which are not managed by CAPI on the ManagedCluster directly. Users should be fairly familiar
+with this pattern since it is already used extensively throughout CAPI, e.g. by modifying a virtual machine
+through its InfraMachine resource for parameters not defined on the Machine resource.
+
+This approach has two key benefits. First, it can leverage ASO's conversion webhooks to allow CAPZ to interact
+with the ManagedCluster through one version of the API, and users to use a different API version, including
+newer or preview (https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2625) API versions.
+Second, since ASO can [adopt existing Azure
+resources](https://azure.github.io/azure-service-operator/guide/frequently-asked-questions/#how-can-i-import-existing-azure-resources-into-aso),
+adopting existing AKS clusters into CAPZ
+(https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1173) additionally would only require
+the extra steps to create the AzureManagedControlPlane referring to the adopted ManagedCluster.
+
+One other consideration is that this approach trades the requirement that CAPZ has the necessary RBAC
+permissions to create ManagedClusters for users having that same capability. CAPZ would still require
+permissions to read, update, and delete ManagedClusters.
+
+The main roadblocks with this method relate to ClusterClass. If CAPZ's AzureManagedControlPlane controller is
```

Comment:
> * One (two? three? 1 + n agent pools) more object to define in the already very lengthy yaml templates

If we're concerned about overall YAML volume, it seems like whether the spec is defined in the ASO resource or the CAPZ resource doesn't make a huge difference. You could probably argue either way whether its better to have more smaller objects or fewer bigger ones.

> * This would result in inconsistent UX where the user now directly creates and updates some ASO CRs (agent pools, managed clusters) but not others (vnets, resource groups) which could be confusing and easily lead to users doing the "wrong" thing

I was almost wondering if whatever pattern we settle on here could be applied more broadly than just the AKS resources (even though that's a non-goal of this proposal).

> this creates a situation where the CAPZ controller and the user could be overriding each other 

I suppose this isn't really any different from today, where a user could update the replica count of an ASO agent pool which would be overridden by CAPZ eventually. But if users are expected to edit some fields in the ASO resource but not others, then that might be an easier mistake to make than it is now. Is there a way users could derive what they should not change based on `managedFields` or by using server-side apply?

> and CAPZ is having to update the spec of a resource it didn't create.

What are the specific issues with this?

Reply:
> If we're concerned about overall YAML volume, it seems like whether the spec is defined in the ASO resource or the CAPZ resource doesn't make a huge difference.

It's not so much about YAML volume but about increased learning curve and complexity of the yaml manifest making it even harder to digest by a human than it is today.

> You could probably argue either way whether its better to have more smaller objects or fewer bigger ones.

I would disagree on this, by having the user compose the cluster themselves (e.g. knowing exactly what objects they need to create), we are pushing even more of the complexity to the users, as opposed to having CAPZ be a layer of abstraction between Cluster API and Azure, i.e. CAPZ translates the concept of a "Cluster" to multiple Azure resources. By making this change CAPZ is no longer doing that translation, the user is. This makes CAPZ less consistent with other CAPI providers and requires users of other providers to learn about ASO and its resources instead of keeping it an implementation detail.

> I was almost wondering if whatever pattern we settle on here could be applied more broadly than just the AKS resources (even though that's a non-goal of this proposal).

Then it's worth calling out as a risk of temporary inconsistency, with a potential mitigation to apply the pattern to other resources in the future (although that would make the above potentially worse)

> What are the specific issues with this?

Shared ownership of the spec could lead to race conditions and overwriting user changes + an inconsistent user experience. Let's say as a user I want to update my AKS agent pool to enable feature x and I also want to update the replica count. With option 1, I would need to go enable feature x on the ASO object for the agent pool directly, but the replica count on the MachinePool object. If I update the replica count field on the ASO agent pool resource directly, it would get overwritten by the controller (since the CAPZ controller manages the replica count). This is different from today because there are no redundant fields in the CAPZ objects that the user creates (AzureMachine pool has no replica count in Spec for example).

Reply:
Makes sense, reflected this in the doc.